<a href="https://colab.research.google.com/github/campospluiza/procesamientoDelHabla/blob/main/CAMPOS_Desafio_Clasificacion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importacion de las librerias necesarias

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier

# Importamos el dataset

Aca fue necesario ajustar los parametros para cargar el dataset ya que tenia un error por encoding. Elegi tambien cargar solamente las columnas necesarias

In [2]:
columns = ['id', 'category', 'title', 'amenities', 'bathrooms', 'bedrooms', 'price', 'square_feet', 'cityname', 'state']

df = pd.read_csv('/content/apartments_for_rent_classified_10K.csv', encoding='latin-1', usecols=columns, delimiter=';', on_bad_lines='skip')

df.head(10)

id                category  \
0  5668626895  housing/rent/apartment   
1  5664597177  housing/rent/apartment   
2  5668626833  housing/rent/apartment   
3  5659918074  housing/rent/apartment   
4  5668626759  housing/rent/apartment   
5  5667891676  housing/rent/apartment   
6  5668627426  housing/rent/apartment   
7  5668626687  housing/rent/apartment   
8  5668610290  housing/rent/apartment   
9  5668627023  housing/rent/apartment   

                                               title  \
0  Studio apartment 2nd St NE, Uhland Terrace NE,...   
1                  Studio apartment 814 Schutte Road   
2  Studio apartment N Scott St, 14th St N, Arling...   
3                     Studio apartment 1717 12th Ave   
4  Studio apartment Washington Blvd, N Cleveland ...   
5                          0 BR in New York NY 10019   
6                   Studio apartment 2432 Penmar Ave   
7  Studio apartment Oak St NW, 16th St NW, Washin...   
8                       Studio apartment 333 Hyde St   
9   Studio apartment A St SE, 19th St SE, Washington   

                                     amenities  bathrooms  bedrooms  price  \
0                                          NaN        NaN       0.0    790   
1                                          NaN        NaN       1.0    425   
2                                          NaN        1.0       0.0   1390   
3                                          NaN        1.0       0.0    925   
4                                          NaN        NaN       0.0    880   
5  Dishwasher,Elevator,Patio/Deck,Pool,Storage        1.0       0.0   2475   
6                                          NaN        NaN       0.0   1800   
7                                          NaN        NaN       0.0    840   
8                                 Refrigerator        1.0       0.0   1495   
9                                          NaN        NaN       0.0    890   

   square_feet       cityname state  
0          101     Washington    DC  
1          106     Evansville    IN  
2          107      Arlington    VA  
3          116        Seattle    WA  
4          125      Arlington    VA  
5          130      Manhattan    NY  
6          132         Venice    CA  
7          136     Washington    DC  
8          138  San Francisco    CA  
9          141     Washington    DC

# Preprocesamiento del dataset y explicacion del objetivo

## Relleno de valores nulos
## Crear variable de clasificación binaria en base a la mediana del precio
## Separación de variables predictoras y objetivo

El objetivo sera clasificar los departamentos en caros o baratos. caros cuestan mas de 1000 dolares, baratos cuestan menos que eso

In [3]:
# Llenar valores nulos
df['amenities'].fillna('None', inplace=True)
df['bathrooms'].fillna(df['bathrooms'].median(), inplace=True)
df['bedrooms'].fillna(df['bedrooms'].median(), inplace=True)

<ipython-input-3-0f75fb84626e>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['amenities'].fillna('None', inplace=True)
<ipython-input-3-0f75fb84626e>:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.

In [4]:
df.head(10)

id                category  \
0  5668626895  housing/rent/apartment   
1  5664597177  housing/rent/apartment   
2  5668626833  housing/rent/apartment   
3  5659918074  housing/rent/apartment   
4  5668626759  housing/rent/apartment   
5  5667891676  housing/rent/apartment   
6  5668627426  housing/rent/apartment   
7  5668626687  housing/rent/apartment   
8  5668610290  housing/rent/apartment   
9  5668627023  housing/rent/apartment   

                                               title  \
0  Studio apartment 2nd St NE, Uhland Terrace NE,...   
1                  Studio apartment 814 Schutte Road   
2  Studio apartment N Scott St, 14th St N, Arling...   
3                     Studio apartment 1717 12th Ave   
4  Studio apartment Washington Blvd, N Cleveland ...   
5                          0 BR in New York NY 10019   
6                   Studio apartment 2432 Penmar Ave   
7  Studio apartment Oak St NW, 16th St NW, Washin...   
8                       Studio apartment 333 Hyde St   
9   Studio apartment A St SE, 19th St SE, Washington   

                                     amenities  bathrooms  bedrooms  price  \
0                                         None        1.0       0.0    790   
1                                         None        1.0       1.0    425   
2                                         None        1.0       0.0   1390   
3                                         None        1.0       0.0    925   
4                                         None        1.0       0.0    880   
5  Dishwasher,Elevator,Patio/Deck,Pool,Storage        1.0       0.0   2475   
6                                         None        1.0       0.0   1800   
7                                         None        1.0       0.0    840   
8                                 Refrigerator        1.0       0.0   1495   
9                                         None        1.0       0.0    890   

   square_feet       cityname state  
0          101     Washington    DC  
1          106     Evansville    IN  
2          107      Arlington    VA  
3          116        Seattle    WA  
4          125      Arlington    VA  
5          130      Manhattan    NY  
6          132         Venice    CA  
7          136     Washington    DC  
8          138  San Francisco    CA  
9          141     Washington    DC

# Variable Objetivo

La variable objetivo en este modelo de clasificación es price_category, una variable binaria que categoriza los apartamentos en dos grupos: "barato" y "caro". Esta variable se define según el valor de la columna price en relación con la mediana de precios de todos los apartamentos en el dataset. Si el price de un apartamento está por encima de la mediana, se le asigna el valor 1 (caro); si está por debajo o es igual a la mediana, se le asigna el valor 0 (barato).

In [5]:
df['price_category'] = (df['price'] > df['price'].median()).astype(int)


df.head(10)

id                category  \
0  5668626895  housing/rent/apartment   
1  5664597177  housing/rent/apartment   
2  5668626833  housing/rent/apartment   
3  5659918074  housing/rent/apartment   
4  5668626759  housing/rent/apartment   
5  5667891676  housing/rent/apartment   
6  5668627426  housing/rent/apartment   
7  5668626687  housing/rent/apartment   
8  5668610290  housing/rent/apartment   
9  5668627023  housing/rent/apartment   

                                               title  \
0  Studio apartment 2nd St NE, Uhland Terrace NE,...   
1                  Studio apartment 814 Schutte Road   
2  Studio apartment N Scott St, 14th St N, Arling...   
3                     Studio apartment 1717 12th Ave   
4  Studio apartment Washington Blvd, N Cleveland ...   
5                          0 BR in New York NY 10019   
6                   Studio apartment 2432 Penmar Ave   
7  Studio apartment Oak St NW, 16th St NW, Washin...   
8                       Studio apartment 333 Hyde St   
9   Studio apartment A St SE, 19th St SE, Washington   

                                     amenities  bathrooms  bedrooms  price  \
0                                         None        1.0       0.0    790   
1                                         None        1.0       1.0    425   
2                                         None        1.0       0.0   1390   
3                                         None        1.0       0.0    925   
4                                         None        1.0       0.0    880   
5  Dishwasher,Elevator,Patio/Deck,Pool,Storage        1.0       0.0   2475   
6                                         None        1.0       0.0   1800   
7                                         None        1.0       0.0    840   
8                                 Refrigerator        1.0       0.0   1495   
9                                         None        1.0       0.0    890   

   square_feet       cityname state  price_category  
0          101     Washington    DC               0  
1          106     Evansville    IN               0  
2          107      Arlington    VA               1  
3          116        Seattle    WA               0  
4          125      Arlington    VA               0  
5          130      Manhattan    NY               1  
6          132         Venice    CA               1  
7          136     Washington    DC               0  
8          138  San Francisco    CA               1  
9          141     Washington    DC               0

In [6]:
X = df[['amenities', 'bathrooms', 'bedrooms', 'square_feet', 'cityname', 'state']]
y = df['price_category']

In [7]:
X.head(10)

amenities  bathrooms  bedrooms  \
0                                         None        1.0       0.0   
1                                         None        1.0       1.0   
2                                         None        1.0       0.0   
3                                         None        1.0       0.0   
4                                         None        1.0       0.0   
5  Dishwasher,Elevator,Patio/Deck,Pool,Storage        1.0       0.0   
6                                         None        1.0       0.0   
7                                         None        1.0       0.0   
8                                 Refrigerator        1.0       0.0   
9                                         None        1.0       0.0   

   square_feet       cityname state  
0          101     Washington    DC  
1          106     Evansville    IN  
2          107      Arlington    VA  
3          116        Seattle    WA  
4          125      Arlington    VA  
5          130      Manhattan    NY  
6          132         Venice    CA  
7          136     Washington    DC  
8          138  San Francisco    CA  
9          141     Washington    DC

In [8]:
y.head(10)

0    0
1    0
2    1
3    0
4    0
5    1
6    1
7    0
8    1
9    0
Name: price_category, dtype: int64

# Preparacion de Datos

### Para preparar los datos, codifica variables categóricas y escala las numéricas.

In [9]:
categorical_features = ['amenities', 'cityname', 'state']
numerical_features = ['bathrooms', 'bedrooms', 'square_feet']

# Transformación y escalado de características
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numerical_features),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_features)
    ])

In [10]:
preprocessor

ColumnTransformer(transformers=[('num', StandardScaler(),
                                 ['bathrooms', 'bedrooms', 'square_feet']),
                                ('cat', OneHotEncoder(handle_unknown='ignore'),
                                 ['amenities', 'cityname', 'state'])])

# Creación del Pipeline y Modelo
### Aquí usaremos un modelo de bosque aleatorio, que se adapta bien a este tipo de datos.

In [11]:
# Crear pipeline con preprocesamiento y modelo
model = Pipeline(steps=[('preprocessor', preprocessor),
                        ('classifier', RandomForestClassifier(random_state=42))])

In [12]:
# División del dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [13]:
X_train

amenities  bathrooms  bedrooms  \
9254       Dishwasher,Fireplace,Patio/Deck,Refrigerator        2.0       3.0   
1561                                               None        1.0       1.0   
1670                   Gym,Patio/Deck,Pool,Washer Dryer        1.0       1.0   
6087  Clubhouse,Dishwasher,Fireplace,Gym,Hot Tub,Par...        1.5       2.0   
6669                                               None        2.0       2.0   
...                                                 ...        ...       ...   
5734                      Clubhouse,Gym,Patio/Deck,Pool        1.0       2.0   
5191                                               None        1.0       1.0   
5390                                    Parking,Storage        1.0       2.0   
860   AC,Clubhouse,Dishwasher,Gated,Gym,Internet Acc...        1.0       2.0   
7270                                               None        2.0       2.0   

      square_feet     cityname state  
9254         1657   Snellville    GA  
1561          566  Minneapolis    MN  
1670          579       Austin    TX  
6087          906     Lynnwood    WA  
6669          979        Fargo    ND  
...           ...          ...   ...  
5734          875   New London    CT  
5191          824     Bellevue    WA  
5390          844  Fond Du Lac    WI  
860           489  Pico Rivera    CA  
7270         1066       Encino    CA  

[8000 rows x 6 columns]

In [14]:
X_test

amenities  bathrooms  bedrooms  \
6252    Clubhouse,Gym,Parking,Patio/Deck,Storage,Tennis        2.0       2.0   
4684                                               None        1.0       2.0   
1731  Cable or Satellite,Dishwasher,Garbage Disposal...        1.0       2.0   
4742  Fireplace,Garbage Disposal,Gym,Patio/Deck,Pool...        1.0       1.0   
4521                                               None        1.0       1.0   
...                                                 ...        ...       ...   
6412                                               None        1.5       1.0   
8285                                               None        2.0       2.0   
7853                               Gym,Pool,Wood Floors        2.0       2.0   
1095  Alarm,Cable or Satellite,Dishwasher,Garbage Di...        1.0       1.0   
6929                                               None        1.5       2.0   

      square_feet          cityname state  
6252          930            Durham    NC  
4684          786     New Brunswick    NJ  
1731          585        Cincinnati    OH  
4742          790            Edmond    OK  
4521          775         Montclair    NJ  
...           ...               ...   ...  
6412          950         Galveston    TX  
8285         1245             Brick    NJ  
7853         1154  Colorado Springs    CO  
1095          515           Houston    TX  
6929         1008         Schofield    WI  

[2000 rows x 6 columns]

In [15]:
y_train

9254    1
1561    0
1670    1
6087    1
6669    1
       ..
5734    0
5191    1
5390    0
860     1
7270    1
Name: price_category, Length: 8000, dtype: int64

In [16]:
y_test

6252    0
4684    1
1731    0
4742    0
4521    1
       ..
6412    1
8285    1
7853    1
1095    0
6929    0
Name: price_category, Length: 2000, dtype: int64

# Entrenamiento del modelo

In [17]:
model.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num', StandardScaler(),
                                                  ['bathrooms', 'bedrooms',
                                                   'square_feet']),
                                                 ('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['amenities', 'cityname',
                                                   'state'])])),
                ('classifier', RandomForestClassifier(random_state=42))])

# Evaluacion del Modelo

In [18]:
y_pred = model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.85      0.83      0.84      1047
           1       0.82      0.84      0.83       953

    accuracy                           0.84      2000
   macro avg       0.84      0.84      0.84      2000
weighted avg       0.84      0.84      0.84      2000



# Mejora del Modelo
### juste de Hiperparámetros del Modelo (Grid Search)
### Con placa de video

In [19]:
from sklearn.model_selection import GridSearchCV

In [21]:
# Definir el rango de hiperparámetros para el RandomForestClassifier
param_grid = {
    'classifier__n_estimators': [50, 100, 200],
    'classifier__max_depth': [None, 10, 20, 30],
    'classifier__min_samples_split': [2, 5, 10]
}

# Crear el GridSearchCV con pipeline completo
grid_search = GridSearchCV(model, param_grid, cv=5, scoring='accuracy', n_jobs=-1)

# Ejecutar la búsqueda en el conjunto de entrenamiento
grid_search.fit(X_train, y_train)

# Obtener los mejores parámetros y rendimiento
print("Best parameters found: ", grid_search.best_params_)
print("Best cross-validation score: ", grid_search.best_score_)

# Reentrenar el modelo con los mejores parámetros en el conjunto completo de entrenamiento
best_model = grid_search.best_estimator_

Best parameters found:  {'classifier__max_depth': None, 'classifier__min_samples_split': 5, 'classifier__n_estimators': 200}
Best cross-validation score:  0.845375


In [22]:
# Predicción en el conjunto de prueba con el mejor modelo
y_pred = best_model.predict(X_test)

# Reporte de clasificación con métricas de precisión, recall y F1-score
print("Classification Report for the Tuned Model:")
print(classification_report(y_test, y_pred))

Classification Report for the Tuned Model:
              precision    recall  f1-score   support

           0       0.87      0.84      0.85      1047
           1       0.83      0.86      0.84       953

    accuracy                           0.85      2000
   macro avg       0.85      0.85      0.85      2000
weighted avg       0.85      0.85      0.85      2000



# Otros Modelos
## Gradient Boosting Classifier

In [23]:
from sklearn.ensemble import GradientBoostingClassifier

# Crear el pipeline con preprocesamiento y modelo Gradient Boosting
model_gb = Pipeline(steps=[('preprocessor', preprocessor),
                           ('classifier', GradientBoostingClassifier(random_state=42))])

# Entrenar el modelo en el conjunto de entrenamiento
model_gb.fit(X_train, y_train)

# Evaluar en el conjunto de prueba
y_pred_gb = model_gb.predict(X_test)
print("Classification Report for Gradient Boosting Classifier:")
print(classification_report(y_test, y_pred_gb))

Classification Report for Gradient Boosting Classifier:
              precision    recall  f1-score   support

           0       0.81      0.84      0.82      1047
           1       0.82      0.79      0.80       953

    accuracy                           0.81      2000
   macro avg       0.81      0.81      0.81      2000
weighted avg       0.81      0.81      0.81      2000



## Logistic Regression

In [24]:
from sklearn.linear_model import LogisticRegression

# Crear el pipeline con preprocesamiento y modelo de Regresión Logística
model_lr = Pipeline(steps=[('preprocessor', preprocessor),
                           ('classifier', LogisticRegression(random_state=42, max_iter=1000))])

# Entrenar el modelo en el conjunto de entrenamiento
model_lr.fit(X_train, y_train)

# Evaluar en el conjunto de prueba
y_pred_lr = model_lr.predict(X_test)
print("Classification Report for Logistic Regression:")
print(classification_report(y_test, y_pred_lr))

Classification Report for Logistic Regression:
              precision    recall  f1-score   support

           0       0.85      0.82      0.84      1047
           1       0.81      0.84      0.82       953

    accuracy                           0.83      2000
   macro avg       0.83      0.83      0.83      2000
weighted avg       0.83      0.83      0.83      2000



**5.6 Resultados obtenidos y su análisis (contrastar métricas, matriz de confusión)**

Para analizar los resultados obtenidos, primero vamos a comparar las métricas de precisión, recall y F1-score de los distintos modelos probados. En general, buscamos el modelo que haya logrado un buen equilibrio entre precisión (qué tan bien identifica correctamente cada clase) y recall (qué tan bien recupera todos los casos de cada clase).

1. **Modelo con ajuste de hiperparámetros**: Al ajustar los hiperparámetros de un modelo (un Random Forest), encontramos los mejores resultados. Este modelo alcanzó una precisión de 0.85 y un F1-score de 0.85. Estos valores indican que el modelo es muy sólido para ambas clases y se comporta de forma balanceada. Esto es evidente en las métricas de "macro avg" y "weighted avg", que también alcanzan el 0.85.

2. **Gradient Boosting Classifier**: Este modelo logró una precisión global de 0.81. Aunque es un modelo interesante y con buenos resultados, su F1-score es menor (0.81), lo que indica un rendimiento ligeramente inferior al modelo anterior. Además, presenta una precisión y recall más bajos para la clase 1, lo que sugiere que le cuesta más identificar correctamente todos los casos de esta clase.

3. **Regresión Logística**: La regresión logística arrojó buenos resultados en general, con una precisión de 0.83 y un F1-score de 0.83. Aunque es un modelo más simple, se comporta bastante bien. Sin embargo, al compararlo con el modelo con ajuste de hiperparámetros, vemos que tiene un rendimiento ligeramente menor.

**Conclusión:** Al contrastar las métricas, el modelo con ajuste de hiperparámetros demostró ser el mejor, ya que alcanzó los valores más altos en precisión, recall y F1-score en general, y mantiene un buen balance entre ambas clases. El Gradient Boosting fue el que menos rendimiento obtuvo en comparación a los otros dos modelos.

**Matriz de Confusión:** Si bien no tenemos los valores específicos de la matriz de confusión, las métricas indican que el modelo con ajuste de hiperparámetros tiene una mejor capacidad de diferenciar entre ambas clases, minimizando tanto los falsos positivos como los falsos negativos.

---

**5.7 Análisis de la experiencia, elección de modelo candidato**

La experiencia de entrenamiento y ajuste de hiperparámetros fue clave en este trabajo práctico. A través del ajuste, pudimos mejorar el rendimiento del modelo seleccionado, logrando una precisión del 85%, la más alta entre los modelos evaluados. Fue un trabajo interesante de hacer, pero dificil, que demandó ajustes incluso cambiando el runtime (cambiamos a placa de video) para que se puediera correr sin problemas.

**Modelo Candidato**: En base a los resultados, el modelo con ajuste de hiperparámetros es el candidato ideal para implementar, ya que alcanzó el mejor balance de precisión y recall. Además, el proceso de ajuste de hiperparámetros permitió optimizarlo para obtener el mejor rendimiento posible.